Make `QuarkusBuild` not pollute Gradle's build cache

snazy opened the pull request:
Make `QuarkusBuild` not pollute Gradle's build cache

Improve Quarkus build in Gradle, proper up-to-date mechanisms, right config-source priorities.

## Background / Motivation

Currently the `QuarkusBuild` task implementation adds even large build artifacts and unmodified dependency jars to Gradle's build cache. This pollutes the Gradle build cache quite a lot and therefore lets archived build caches become unnecessary huge.
On top of that, "relicts" from previous runs of `QuarkusBuild` using a different configuration, like a different package-type, causes multiple permutations of the cached artifacts and leads to restoring unexpected artifacts from the build cache.

This large change updates the Gradle logic to have "proper" up-to-date checks and in turn a reasonable setting of what is cached and what is not cached.

Background on how I discovered the "caching and up-to-date issue" with `QuarkusBuild`: Quarkus 2.16 added the `@CacheableTask` annotation to `QuarkusBuild`. This means that _everything_ that is generated by `QuarkusBuild` (and "declared" as an output: fast-jar directory, uber-jar, native-runner) is cached. After having Quarkus 2.16.x in our "main" branch for about 1/2 day, nearly the whole "10 GB budget" for cache artifacts in GitHub was occupied by Gradle build cache artifacts - the build cache artifacts grew from build to build - this was not the case without the `@CacheableTask` annotation.

The current inputs of the `QuarkusBuild` task are also incomplete. Those inputs do not include for example the configured `finalName`, the settings on the Gradle project and the settings on the Quarkus extension. This leads to situations that the wrong Gradle cache artifacts could be restored (think: you ask for a fast-jar, but get an uber-jar).

Another issue is that repeated runs of `quarkusBuild` against the exact same code base and dependencies leads to different outputs, the contents of the generated jars and e.g. quarkus-application.dat differ. This causes different outputs of the `QuarkusBuild` task, which cause depending tasks and projects to unnecessarily run/work again.

## Initial Goal

The initial goal of this effort was to make caching of `fast-jar` builds work nice in CI environments, I assume that's the most common package type used in CI environments for example for testing purposes. The idea here is to _not_ cache all the dependency jars, but leverage the mechanisms available in Gradle to "just collect" the dependencies. In other words (and simplified): for `fast-jar`, cache everthing "in `quarkus-app` except `lib/`.

## Updated Quarkus Gradle build implementation

There is no change for users running a `./gradlew quarkusBuild`. `QuarkusBuild` is still the task that users should execute (or depend on in their builds).

To achive the goal to make especially `fast-jar` builds "nicely cacheable" in CI, two new tasks had to be introduced:

* `QuarkusBuildDependencies` uses the dependency information from the Quarkus `ApplicationModel` to tell Gradle to collect those. This task works for `fast-jar` and `legacy-jar` and is a "no op" for other package types. This task is _never_ cacheable, but has _up-to-date_ checks.
* `QuarkusBuildCacheableAppParts` performs a Quarkus application build and collects the _cacheable_ parts of that build. This task works for `fast-jar` and `legacy-jar` and is a "no op" for other package types. This task is cacheable.
* `QuarkusBuild`, for `fast-jar` and `legacy-jar`, assembles the outputs of the above two tasks to produce the expected output. Performs a Quarkus build for other package types. See below on when this task is cacheable.

The `build/quarkus-build/` directory is used to properly separate the outputs of the above three tasks:

* `build/quarkus-build/gen/` receives the full output of a Quarkus build
* `build/quarkus-build/app/` receives the _cacheable_ parts from `build/quarkus-build/gen/`
* `build/quarkus-build/dep/` receives the dependency jars

## CI vs non-CI

The output of `QuarkusBuild` is, by default, cacheable in non-CI environments and _not_ cacheable in CI environments. The behavior can be explicitly overridden using the new property `cacheLargeArtifacts` property. As outlined above, caching huge artifacts is not beneficial in CI, either due to space limitations (GitHub's 10GB limit for example) or just the cost of network traffic/duration. On a developer's machine however, the cost of storing/retrieving even bigger artifacts is rather neglectible and lower than rebuilding a Quarkus application.

## Configuration

Before this change, the "priority" of configuration settings was (in most cases...) effectively: System properties, environment variables, `application.properties`, configurations in Gradle build scripts & Gradle project properties. This order is unintuitive and was changed to: system properties, environment variables, Gradle project properties, Quarkus build configuration, application.properties.  The previous code had several places in which Quarkus related configuration settings were retrieved with sometimes different priorities of the "config sources".

The whole "machinery" to get the configuration has been rewritten and encapsulated in new classes `EffectiveConfig` (effective for a Quarkus build) and `BaseConfig` (available during Gradle task configuration phase). `EffectiveConfig` is not really different from `BaseConfig, but contains the "special" settings from e.g. `ImagePush` task.

The new implementation also uses SmallRye Config and especially the existing Quarkus mechanisms to pull information out of a `PackageConfig` _object_ produced from a configuration. It turned out to be easier to reason about `PackageConfig` than "raw property values".

Support for `application.yaml/yml` has also been added.

## Workers

All calls into the "inner workings" of a Quarkus build and Quarkus code generation have been moved to separate worker processes. The code in this change to support this is pretty dead simple.

The primary reason to move that work into isolated worker processes is that all the `QuarkusBootstrap` and derived pieces know nothing about Gradle, so there is no "property like" mechanism to override settings from `application.properties/yaml/yml` with those from e.g. the `QuarkusPluginExtension`. The only option would have been to modify the system properties of the Gradle build process - but that's a no-go, especially considering other tasks running in parallel (think: two `QuarkusBuild` of different projects running at the same time). It would have been relatively easy to serialize all `QuarkusBuild` actions across a Gradle build, but why - and it would prevent using "beefy machines" to run many `QuarkusBuild`s in parallel. Another option would have been to implement a rather complex (and likely very racy) mechanism to track modifications to system properties.

As a conclusion, it wasn't just very simple to leverage the process isolation using the Gradle worker API, but it's also not bad. Gradle does reuse already spawned and compatible worker instances.

The "trick" implemented to "prioritize" configs from Gradle project properties and Quarkus extension settings is to pass the whole config as system properties to the worker.

## Logging / cache-disabled messages

A bunch of hopefully useful logging has been implemented and added.

With info-level loogging, the tasks emit at least some basic information about what they do and at least the package type being used.

## Effective config

To be able to investigate which configuration settings were effectively used, there's another new task called `quarkusShowEffectiveConfig` that shows all the `quarkus.*` properties and some more information, including the loaded `application.(properties|yaml|yml)`. This task is intended to debug build issues. The task can optionally save the effecitve config properties as a properties file in the `build/` directory, if run with the command line option `./gradlew quarkusShowEffectiveConfig --save-config-properties`.

## Test changes

None of the existing tests has changed, except the `BuildCOnfigurationTest` had to be adopted to reflect the updated order of config sources.

Relates to: #30852

snazy committed:
Make `QuarkusBuild` not pollute Gradle's build cache

Improve Quarkus build in Gradle, proper up-to-date mechanisms, right config-source priorities.

Currently the `QuarkusBuild` task implementation adds even large build artifacts and unmodified dependency jars to Gradle's build cache. This pollutes the Gradle build cache quite a lot and therefore lets archived build caches become unnecessary huge.
On top of that, "relicts" from previous runs of `QuarkusBuild` using a different configuration, like a different package-type, causes multiple permutations of the cached artifacts and leads to restoring unexpected artifacts from the build cache.

This large change updates the Gradle logic to have "proper" up-to-date checks and in turn a reasonable setting of what is cached and what is not cached.

Background on how I discovered the "caching and up-to-date issue" with `QuarkusBuild`: Quarkus 2.16 added the `@CacheableTask` annotation to `QuarkusBuild`. This means that _everything_ that is generated by `QuarkusBuild` (and "declared" as an output: fast-jar directory, uber-jar, native-runner) is cached. After having Quarkus 2.16.x in our "main" branch for about 1/2 day, nearly the whole "10 GB budget" for cache artifacts in GitHub was occupied by Gradle build cache artifacts - the build cache artifacts grew from build to build - this was not the case without the `@CacheableTask` annotation.

The current inputs of the `QuarkusBuild` task are also incomplete. Those inputs do not include for example the configured `finalName`, the settings on the Gradle project and the settings on the Quarkus extension. This leads to situations that the wrong Gradle cache artifacts could be restored (think: you ask for a fast-jar, but get an uber-jar).

Another issue is that repeated runs of `quarkusBuild` against the exact same code base and dependencies leads to different outputs, the contents of the generated jars and e.g. quarkus-application.dat differ. This causes different outputs of the `QuarkusBuild` task, which cause depending tasks and projects to unnecessarily run/work again.

The initial goal of this effort was to make caching of `fast-jar` builds work nice in CI environments, I assume that's the most common package type used in CI environments for example for testing purposes. The idea here is to _not_ cache all the dependency jars, but leverage the mechanisms available in Gradle to "just collect" the dependencies. In other words (and simplified): for `fast-jar`, cache everthing "in `quarkus-app` except `lib/`.

There is no change for users running a `./gradlew quarkusBuild`. `QuarkusBuild` is still the task that users should execute (or depend on in their builds).

To achive the goal to make especially `fast-jar` builds "nicely cacheable" in CI, two new tasks had to be introduced:

* `QuarkusBuildDependencies` uses the dependency information from the Quarkus `ApplicationModel` to tell Gradle to collect those. This task works for `fast-jar` and `legacy-jar` and is a "no op" for other package types. This task is _never_ cacheable, but has _up-to-date_ checks.
* `QuarkusBuildCacheableAppParts` performs a Quarkus application build and collects the _cacheable_ parts of that build. This task works for `fast-jar` and `legacy-jar` and is a "no op" for other package types. This task is cacheable.
* `QuarkusBuild`, for `fast-jar` and `legacy-jar`, assembles the outputs of the above two tasks to produce the expected output. Performs a Quarkus build for other package types. See below on when this task is cacheable.

The `build/quarkus-build/` directory is used to properly separate the outputs of the above three tasks:

* `build/quarkus-build/gen/` receives the full output of a Quarkus build
* `build/quarkus-build/app/` receives the _cacheable_ parts from `build/quarkus-build/gen/`
* `build/quarkus-build/dep/` receives the dependency jars

The output of `QuarkusBuild` is, by default, cacheable in non-CI environments and _not_ cacheable in CI environments. The behavior can be explicitly overridden using the new property `cacheLargeArtifacts` property. As outlined above, caching huge artifacts is not beneficial in CI, either due to space limitations (GitHub's 10GB limit for example) or just the cost of network traffic/duration. On a developer's machine however, the cost of storing/retrieving even bigger artifacts is rather neglectible and lower than rebuilding a Quarkus application.

Before this change, the "priority" of configuration settings was (in most cases...) effectively: System properties, environment variables, `application.properties`, configurations in Gradle build scripts & Gradle project properties. This order is unintuitive and was changed to: system properties, environment variables, Gradle project properties, Quarkus build configuration, application.properties.  The previous code had several places in which Quarkus related configuration settings were retrieved with sometimes different priorities of the "config sources".

The whole "machinery" to get the configuration has been rewritten and encapsulated in new classes `EffectiveConfig` (effective for a Quarkus build) and `BaseConfig` (available during Gradle task configuration phase). `EffectiveConfig` is not really different from `BaseConfig, but contains the "special" settings from e.g. `ImagePush` task.

The new implementation also uses SmallRye Config and especially the existing Quarkus mechanisms to pull information out of a `PackageConfig` _object_ produced from a configuration. It turned out to be easier to reason about `PackageConfig` than "raw property values".

Support for `application.yaml/yml` has also been added.

All calls into the "inner workings" of a Quarkus build and Quarkus code generation have been moved to separate worker processes. The code in this change to support this is pretty dead simple.

The primary reason to move that work into isolated worker processes is that all the `QuarkusBootstrap` and derived pieces know nothing about Gradle, so there is no "property like" mechanism to override settings from `application.properties/yaml/yml` with those from e.g. the `QuarkusPluginExtension`. The only option would have been to modify the system properties of the Gradle build process - but that's a no-go, especially considering other tasks running in parallel (think: two `QuarkusBuild` of different projects running at the same time). It would have been relatively easy to serialize all `QuarkusBuild` actions across a Gradle build, but why - and it would prevent using "beefy machines" to run many `QuarkusBuild`s in parallel. Another option would have been to implement a rather complex (and likely very racy) mechanism to track modifications to system properties.

As a conclusion, it wasn't just very simple to leverage the process isolation using the Gradle worker API, but it's also not bad. Gradle does reuse already spawned and compatible worker instances.

The "trick" implemented to "prioritize" configs from Gradle project properties and Quarkus extension settings is to pass the whole config as system properties to the worker.

A bunch of hopefully useful logging has been implemented and added.

With info-level loogging, the tasks emit at least some basic information about what they do and at least the package type being used.

To be able to investigate which configuration settings were effectively used, there's another new task called `quarkusShowEffectiveConfig` that shows all the `quarkus.*` properties and some more information, including the loaded `application.(properties|yaml|yml)`. This task is intended to debug build issues. The task can optionally save the effecitve config properties as a properties file in the `build/` directory, if run with the command line option `./gradlew quarkusShowEffectiveConfig --save-config-properties`.

None of the existing tests has changed, except the `BuildCOnfigurationTest` had to be adopted to reflect the updated order of config sources.

Relates to: #30852

aloubyansky commented on integration-tests/gradle/src/test/java/io/quarkus/gradle/FastJarFormatWorksTest.java:
```diff
@@ -46,22 +46,9 @@ public void testFastJarFormatWorks() throws Exception {
-            // test that the application name and version are properly set
-            assertThat(DevModeTestUtils.getHttpResponse("/hello", () -> {
-                return jarRun == null ? null : jarRun.isDone() ? "jar run mode has terminated" : null;
-            }).equals("hello"));
```
What was the reason to remove the endpoint check? Looks like this is what is actually verifying that the app responds.

snazy replied:
re-added (earlier)

aloubyansky commented on devtools/gradle/gradle-application-plugin/build.gradle:
```diff
@@ -7,6 +7,7 @@ dependencies {
     implementation "io.quarkus:quarkus-devtools-common:${version}"
     implementation "io.quarkus:quarkus-core-deployment:${version}"
     implementation "io.quarkus:quarkus-bootstrap-gradle-resolver:${version}"
+    implementation "io.smallrye.config:smallrye-config-source-yaml:${smallrye_config_version}"
```
Looks like we should be enforcing the `quarkus-bom`, that's not for this PR though.

aloubyansky commented on devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/QuarkusPlugin.java:
```diff
@@ -187,7 +210,6 @@ private void registerTasks(Project project, QuarkusPluginExtension quarkusExt) {
 
         });
 
-        configureBuildNativeTask(project);
```
What happens now if BUILD_NATIVE_TASK_NAME is executed?

snazy replied:
Hm - wonder whether it's okay to remove both `buildNative` + `testNative`, as those are already deprecated?

aloubyansky replied:
While it's deprecated, I suspect people will ask for it if we remove it. If I'm not mistaken it happened before. Could we keep it for now?

snazy replied:
Re-added it and adopted the native-integration-tests a little (not sure why those didn't fail).

aloubyansky commented on devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/extension/QuarkusPluginExtension.java:
```diff
@@ -25,31 +24,44 @@
+        this.cleanupBuildOutput = project.getObjects().property(Boolean.class)
+                .convention(true);
+        this.cacheLargeArtifacts = project.getObjects().property(Boolean.class)
+                .convention(!System.getenv().containsKey("CI"));
```
Is this a common/standard env var?

snazy replied:
Yes

aloubyansky commented on devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/EffectiveConfig.java:
```diff
@@ -0,0 +1,219 @@
+        // 500 -> System.getenv()
+        // 480 -> quarkusBuildProperties
+        // 300 -> projectProperties
+        // 0-250 -> application.(properties|yaml|yml)
```
@radcortez could you please review the config composition? Thanks.

snazy replied:
Didn't knew that list.
Updated the ordinals + added microprofile.properties as well.

aloubyansky commented:
FYI @glefloch @jskillin-idt

jskillin-idt commented:
> FYI @glefloch @jskillin-idt

I appreciate the ping! :+1:

snazy commented:
Tried to get an understanding of the Windows failures - the assertions check the exit-code to be 0, and the Gradle builds in the ITs finished successfully - no clue why that fails (and no Windows machine here).

aloubyansky commented:
There is something wrong in the Gradle Windows CI in other PRs too.

snazy commented:
> Although at least the gRPC tests seem to pass in other PRs.

Strange though, still cannot repro it. :(

aloubyansky commented:
@snazy could you disable those tests on windows?

snazy commented:
> @snazy could you disable those tests on windows?

Sure, done + rebased

snazy commented:
Hum - so that's not nice :(

Finally found some hint in the logs:
```
> Task :quarkusGenerateCode
Generating Quarkus code for org.test:test-codestart:1.0.0-codestart
  launch mode:                  NORMAL
  base name:                    test-codestart-1.0.0-codestart
  generated source directory:   D:\a\quarkus\quarkus\integration-tests\devtools\target\quarkus-codestart-build-test\project-gradle-kotlin-dsl-java-d770bd4b-e0d6-4c8b-a697-ebe48232953a\build\classes\java\quarkus-generated-sources
  build directory:              D:\a\quarkus\quarkus\integration-tests\devtools\target\quarkus-codestart-build-test\project-gradle-kotlin-dsl-java-d770bd4b-e0d6-4c8b-a697-ebe48232953a\build
'"java.exe"' is not recognized as an internal or external command,

operable program or batch file.
--q-grpc_out: protoc-gen-q-grpc: Plugin failed with status code 1.


FAILURE: Build failed with an exception.
> Task :quarkusGenerateCode FAILED
```

I've reverted the test-ignores for Windows, added the `--stracktrace` argument to some more Gradle invocations and (temporarily) added unconditional logging of the error stacktrace.

The error above comes from some code invoked by `io.quarkus.gradle.tasks.worker.CodeGenWorker`. Not sure whether the double _and_ single quotes around `java.exe` are some hint though.

snazy committed:
Unify Gradle process worker configuration

snazy committed:
Hack-ish fix for Windows, generating gRPC

snazy commented:
For posterity: some gRPC/codegen integration tests fail(ed) only on Windows, so likely an OS specific issue, but not 100% clear where and why exactly. On a "fresh" Windows machine with just a Java + Git installed, and Java w/o all the environments like `JAVA_HOME`, `PATH` and w/o the registry keys configured, the tests failing in Windows-CI failed on that machine as well. Some results from experiments:
* `JAVA_HOME` is mandatory for both Maven + Gradle to find the Java executable (not really a surprise)
* Having `JAVA_HOME\bin` in `PATH` is neither necessary for Maven nor for Gradle (also not really a surprise). But it **is** necessary for at least the grpc code generation, likely the codegen binaries.
* Populating both `JAVA_HOME` + `PATH` to the worker processes seems to fix it - as implemented in `GradleTask.configureProcessWorkerSpec()`.

gsmet commented:
Thanks @snazy , hopefully, this won't break the Internet :).
